### PR TITLE
Closes #1452: Implements auto-generation of custom recipes

### LINKED ISSUE #1452: Auto-Generate custom recipes
If custom recipes is set to `true` it expects the user to copy the custom recipes into the config folder. This is rarely applicable, because it requires usage-knowledge.

Auto-Exporting will highly improve the QoL for custom recipes user.

The should copy all files named with `*.recipe` in a specific path to the config location. This is extendable for having `*.recipe` files all over the package system if required.

### PR DESCRIPTION
**Pull request by thatsIch**
Removes config option for `B:customRecipes`

It now generates two folders in your config folder:

* `generated recipes`
* `user recipes`

the entry point is still `index.recipes` but if a replacement can be found in the `user recipes` these are prioritized. That way you do not need to replace all recipe files, but only the one you want to replace.

### CODE REVIEW

**Comment by yueh** [comment]
This actually has some pretty serious flaw as it will prevent us from updating any recipe after the user has started it the first time and I doubt most players will manually enforce a reset every time they update AE2.

Maybe add two folders for the recipe to the config folder? One for the default recipes, which will be overwritten with every start (or store the version somewhere and only overwrite with an update) and a second one for the actual custom recipes.

**Comment by thatsIch** [comment]
Self modified recipes are always prioritized, 
this does not change.

**Comment by thatsIch** [comment]
thought about `generated recipes` and `user recipes` folders.
Probably need somehow to highlight the `index` somehow, by either pushing the rest up one higher or something along these lines.

**Comment by yueh** [comment]
The recipes should simultaneously receive an overhaul themselves and be better organized.

**Comment by thatsIch** [comment]
I thought about removing the ´index` and have for each recipe a single file. 
That would make priority recipes maybe easier.

**Comment by yueh** [comment]
One file per recipe might get really tedious to maintain for someone actually wanting to change recipes.
Maybe group them a bit? Like one file for all dense cables (colors), one for upgrade cards, on for QNB related and so on?

And then there is still the question on where to place new recipes. Like crafting processors without inscribers, or recipes for presses.

**Comment by thatsIch** [comment]
Updated description

**Comment by thatsIch** [comment]
This PR handles the auto generation of recipes. There is another issue open for the flattening and the general rework of recipes.

File: src/main/java/appeng/core/RecipeLoader.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+		}
+		// on failure use jar parsing
+		catch( IOException e )
+		{
+			e.printStackTrace();
+			this.handler.parseRecipes( new JarLoader( "/assets/appliedenergistics2/recipes/" ), "index.recipe" );
+		}
+		catch( URISyntaxException e )
+		{
+			e.printStackTrace();
```
Comment (yueh) [inline src/main/java/appeng/core/RecipeLoader.java]: Should probably be changed to a more meaningful `AELog`. Else people will start complaining about "stacktraces are crashing my game" :grin:

File: src/main/java/appeng/core/RecipeLoader.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+	}
+
+	@Override
+	public void run()
+	{
+		// setup copying
+		final RecipeResourceCopier copier = new RecipeResourceCopier( "assets/appliedenergistics2/recipes/" );
+		final File configDirectory = AppEng.instance().getConfigDirectory();
+		final File generatedRecipesDir = new File( configDirectory, "generated recipes" );
+		final File userRecipesDir = new File( configDirectory, "user recipes" );
```
Comment (yueh) [inline src/main/java/appeng/core/RecipeLoader.java]: We should probably replace the spaces with something else. For example traversing the folders on linux with a console is slightly annoying as spaces needs to be escaped. So either something like `userRecipes` or `user-recipes` is a bit more convenient.

**Comment by yueh** [comment]
It might also be a good idea to add a readme to each folder.

For example to warn them that the `generated recipes` will be replaced and changes are then lost as well as some information to copy the file over to `user recipes` as well as a link to [custom recipes](http://ae2.ae-mod.info/Custom-Recipes/) for further information.

File: src/main/java/appeng/core/RecipeLoader.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+			copier.copyTo( generatedRecipesDir );
+
+			// parse recipes prioritising the user scripts by using the generated as template
+			this.handler.parseRecipes( new ConfigLoader( generatedRecipesDir, userRecipesDir ), "index.recipe" );
+		}
+		// on failure use jar parsing
+		catch( IOException e )
+		{
+			AELog.error( e );
```
Comment (yueh) [inline src/main/java/appeng/core/RecipeLoader.java]: `AElog.error()` itself does a `e.printStackTrace()` and an additional line with the exception class + message. Maybe we should also the stacktrace there.

Comment (thatsIch) [inline src/main/java/appeng/core/RecipeLoader.java]: verb missing xD

**Comment by thatsIch** [comment]
Added readme and copying it over to `generated` and `user` directory. 
Should always be a fresh readme from JAR / FS

**Pull request merged**